pubmed-semantic.py

```diff
@@ -0,0 +1,250 @@
+    PubmedArticleSet: PubmedArticleSet
+
+
+# -- NCBI EUTILS --
+def _safe_request(url: str, method: str = "GET", headers={}, **opts):
+    user_agent = f"{settings.app_name}/{settings.app_version} ({settings.app_url};mailto:{settings.admin_email})"
+    request_headers = {"user-agent": user_agent}
+    request_headers.update(headers)
+    try:
+        r = requests.request(
+            method, url, headers=request_headers, timeout=settings.http_request_timeout, **opts
+        )
+        r.raise_for_status()
+    except requests.exceptions.Timeout as e:
+        print(f"Timeout error {e}")
+        raise
+    except requests.exceptions.HTTPError as e:
+        print(f"HTTP error {e}; status code: {r.status_code}")
+        raise
+    except requests.exceptions.RequestException as e:
+        print(f"Error in request {e}")
+        raise
+    else:
+        return r
+
+
+def _get_eutil_records(eutil: str, id: List[str], **opts) -> dict:
+    """Call one of the NCBI EUTILITIES and returns data as Python objects."""
+    eutils_params = {
+        "db": "pubmed",
+        "id": ",".join(id),
+        "retstart": 0,
+        "retmode": "xml",
+        "api_key": settings.ncbi_eutils_api_key,
+    }
+    eutils_params.update(opts)
+    if eutil == "esummary":
+        url = settings.eutils_esummary_url
+    elif eutil == "efetch":
+        url = settings.eutils_efetch_url
+    else:
+        raise ValueError(f"Unsupported eutil '{eutil}''")
+    eutilResponse = _safe_request(url, "POST", files=eutils_params)
+    doc = xmltodict.parse(eutilResponse.text)
+    return doc
+
+
+def _articles_to_docs(articles: List[PubmedArticle]) -> List[Dict[str, Collection[Any]]]:
+    """Return a list Documents given a PubmedArticle"""
+    docs = []
+    for pubmed_article in articles:
+        abstract = pubmed_article.MedlineCitation.Article.Abstract
+        title = pubmed_article.MedlineCitation.Article.ArticleTitle
+        text = " ".join(_compact([title, abstract]))
+        pmid = pubmed_article.MedlineCitation.PMID
+        docs.append({"uid": pmid, "text": text})
+    return docs
+
+
+# -- Public methods --
+def uids_to_docs(uids: List[str]) -> List[Dict[str, Collection[Any]]]:
+    """Return uid, and text (i.e. title + abstract) given a PubMed uid"""
+    docs: List[Dict[str, Collection[Any]]] = []
+    num_uids = len(uids)
+    num_queries = num_uids // MAX_EFETCH_RETMAX + 1
+    for i in range(num_queries):
+        lower = i * MAX_EFETCH_RETMAX
+        upper = min([lower + MAX_EFETCH_RETMAX, num_uids])
+        id = uids[lower:upper]
+        try:
+            eutil_response = _get_eutil_records("efetch", id)
+            ERROR = _get(eutil_response, ["eFetchResult", "ERROR"])
+            if ERROR:
+                raise RuntimeError(ERROR)
+            pubmedEfetchReponse = PubmedEfetchResponse(**eutil_response)
+            pubmedArticle = pubmedEfetchReponse.PubmedArticleSet.PubmedArticle
+        except Exception as e:

```

Comment:
Just FYI: When something fails validation, it throws a [`ValidationError` ](https://pydantic-docs.helpmanual.io/usage/validators/) that you can handle. Currently this just kills the process, but you could handle it here (e.g. track the ids you missed) and continue the loop.